File: backoffice/src/diningTable/application/DiningTableNumberChanger.py
"""
 *
 * Libraries 
 *
"""

from src.restaurant.domain  import RestaurantId
from src.diningTable.domain import DiningTableRepository
from src.shared.domain      import DomainEventsPublisher
from src.diningTable.domain import DiningTable
from src.diningTable.domain import DiningTableId
from src.diningTable.domain import DiningTableNumber
from src.shared.domain      import ServerInternalErrorException
from src.diningTable.domain import DiningTableNotFoundException
from src.diningTable.domain import DiningTableNumberAlreadyCreatedException

"""
 *
 * Class
 *
"""

class DiningTableNumberChanger:

    """
     *
     * Parameters 
     *
    """
 
    __repository     : DiningTableRepository
    __eventPublisher : DomainEventsPublisher

    """
     *
     * Methods 
     *
    """

    def __init__( 
        self, 
        repository     : DiningTableRepository,
        eventPublisher : DomainEventsPublisher
    ) -> None:
        self.__repository     = repository
        self.__eventPublisher = eventPublisher
    
    def change(
        self, 
        id           : DiningTableId,
        number       : DiningTableNumber,
        restaurantId : RestaurantId
    ) -> None:
        # Variables
        diningTable : DiningTable
        # Code
        diningTable = self.__repository.findByNumberAndRestaurant( number, restaurantId )
        if diningTable is not None:
            raise DiningTableNumberAlreadyCreatedException( number )
        diningTable = self.__repository.findByIdAndRestaurant( id, restaurantId )
        if diningTable is None:
            raise DiningTableNotFoundException( id )
        diningTable.renumber( number )
        if not self.__repository.update( diningTable ):
            raise ServerInternalErrorException()
        self.__eventPublisher.publish( diningTable.pullEvents() )

File: preparation/src/diningTable/application/DiningTableNumberChanger.cs
/* 
 * 
 * Libraries
 *
*/

using preparation.src.diningTable.domain;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.diningTable.application {

    public class DiningTableNumberChanger : DomainEventSubscriber<DiningTableNumberChanged> {

        /* 
         * 
         * Attributes
         *
        */

        private DiningTableRepository repository;

        /* 
         * 
         * Methods
         *
        */

        public DiningTableNumberChanger( DiningTableRepository repository ) {
            this.repository = repository;
        }

        public async Task handleEvent( DiningTableNumberChanged domainEvent ) {
            // Variables
            DiningTable? diningTable;
            // Code
            await Task.Run( () => {
                diningTable = repository.findById( domainEvent.getId() );
                if( diningTable == null ) {
                    throw new DiningTableNotFoundException( domainEvent.getId() );                    
                }
                diningTable.changeNumber( domainEvent.getNumber() );
                if( !repository.update( diningTable ) ) {
                    throw new ServerInternalErrorException();
                }
            } );
        }

    }

}